lib/SPIRV/libSPIRV/SPIRVAsm.h

```diff
@@ -0,0 +1,155 @@
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+/// \file
+///
+/// This file defines the inline assembler entries defined in SPIRV spec with op
+/// codes.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SPIRV_LIBSPIRV_SPIRVASM_H
+#define SPIRV_LIBSPIRV_SPIRVASM_H
+
+#include "SPIRVEntry.h"
+#include "SPIRVInstruction.h"
+#include "SPIRVStream.h"
+
+#include <cassert>
+#include <map>
+#include <tuple>
+#include <vector>
+
+namespace SPIRV {
+
+class SPIRVAsmTargetINTEL : public SPIRVEntry {
+public:
+  static const SPIRVWord FixedWC = 2;
+  static const Op OC = OpAsmTargetINTEL;
+  // Complete constructor
+  SPIRVAsmTargetINTEL(SPIRVModule *M, SPIRVId TheId,
+                      const std::string &TheTarget)
+      : SPIRVEntry(M, FixedWC + getSizeInWords(TheTarget), OC, TheId),
+        Target(TheTarget) {
+    validate();
+  }
+  // Incomplete constructor
+  SPIRVAsmTargetINTEL() : SPIRVEntry(OC) {}
+  SPIRVCapVec getRequiredCapability() const override {
+    return getVec(CapabilityAsmINTEL);
+  }
+  SPIRVExtSet getRequiredExtensions() const override {
+    return getSet(ExtensionID::SPV_INTEL_inline_assembly);
+  }
+  const std::string &getTarget() const { return Target; }
+
+protected:
+  void validate() const override {
+    SPIRVEntry::validate();
+    assert(WordCount > FixedWC);
+    assert(OpCode == OC);
+  }
+  _SPIRV_DEF_ENCDEC2(Id, Target)
+  std::string Target;
+};
+
+class SPIRVAsmINTEL : public SPIRVValue {
+public:
+  static const SPIRVWord FixedWC = 5;
+  static const Op OC = OpAsmINTEL;
+  // Complete constructor
+  SPIRVAsmINTEL(SPIRVModule *M, SPIRVTypeFunction *TheFunctionType,
+                SPIRVId TheId, SPIRVAsmTargetINTEL *TheTarget,
+                const std::string &TheInstructions,
+                const std::string &TheConstraints)
+      : SPIRVValue(M,
+                   FixedWC + getSizeInWords(TheInstructions) +
+                       getSizeInWords(TheConstraints),
+                   OC, TheFunctionType->getReturnType(), TheId),
+        Target(TheTarget), FunctionType(TheFunctionType),
+        Instructions(TheInstructions), Constraints(TheConstraints) {
+    validate();
+  }
+  // Incomplete constructor
+  SPIRVAsmINTEL() : SPIRVValue(OC) {}
+  SPIRVCapVec getRequiredCapability() const override {
+    return getVec(CapabilityAsmINTEL);
+  }
+  SPIRVExtSet getRequiredExtensions() const override {
+    return getSet(ExtensionID::SPV_INTEL_inline_assembly);
+  }
+  const std::string &getInstructions() const { return Instructions; }
+  const std::string &getConstraints() const { return Constraints; }
+  SPIRVTypeFunction *getFunctionType() const { return FunctionType; }
+
+protected:
+  _SPIRV_DEF_ENCDEC6(Type, Id, FunctionType, Target, Instructions, Constraints)
+  void encodeParents(spv_ostream &O) const {
+    O << SPIRVNL();
+    O << *Target;
```

Comment:
I think we should encode `OpAsmTarget` in some other way.

Reply:
Can you be more precise on what is wrong in this way and what do you propose?

Reply:
I mean you don't need it if the code will be changed according to my previous comment https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/454#discussion_r393691770